mutable_buffer/src/table.rs

```
@@ -237,6 +240,106 @@ impl Table {
+                            if !insert_column.is_field() {
+                                InternalColumnTypeMismatch {
+                                    column_id,
+                                    expected_column_type: c.type_description(),
+                                    actual_column_type: values.type_description(),
+                                }
+                                .fail()?
+                            };
+                        }
+                        (entry::TypedValuesIterator::String(_), Column::Tag(_, _)) => {
+                            if !insert_column.is_tag() {
+                                InternalColumnTypeMismatch {
+                                    column_id,
+                                    expected_column_type: c.type_description(),
+                                    actual_column_type: values.type_description(),
+                                }
+                                .fail()?
+                            };
+                        }
+                        _ => InternalColumnTypeMismatch {
+                            column_id,
+                            expected_column_type: c.type_description(),
+                            actual_column_type: values.type_description(),
+                        }
+                        .fail()?,
+                    }
+                }
+
+                Ok((column_id, insert_column.logical_type(), values))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        let row_count_before_insert = self.row_count();
+
+        for (column_id, logical_type, values) in columns_with_inserts.into_iter() {
+            match self.columns.get_mut(&column_id) {
+                Some(c) => c
+                    .push_typed_values(dictionary, logical_type, values)
+                    .with_context(|| {
+                        let column = dictionary.lookup_id(column_id).unwrap_or("unknown");
```

I think using a default value "unknown" would masks a serious invariant breaking. The `column_id` here comes from the column name inserted in the dictionary by this very function. If it's not possible to retrieve it here it means something is seriously wrong and we should crash asap.

---

What @mkmik mentions here is a general theme I worry about when returning "bad values" and not crashing. I don't take issue with this approach but you should have considered the ramifications and blast radius. Crashing can be better than infecting state with the wrong data.

---

Sounds good, I'll replace these defaults with `expect`